bot/slack_im_notifier.go

```diff
@@ -3,26 +3,52 @@ package bot
 import (
 	"fmt"
 	"log"
+	"sync"
+	"time"
 
-	"github.com/andrewslotin/michael/deploy"
-	"github.com/andrewslotin/michael/slack"
+	"github.com/adjust/michaelbot/deploy"
+	"github.com/adjust/michaelbot/slack"
 )
 
 type SlackIMNotifier struct {
-	im    *slack.InstantMessenger
-	users *slack.TeamDirectory
+	im             *slack.InstantMessenger
+	users          *slack.TeamDirectory
+	warningTimeout time.Duration
+	timer          *time.Timer
+	mutex          sync.Mutex
 }
 
-func NewSlackIMNotifier(api *slack.WebAPI) *SlackIMNotifier {
+func NewSlackIMNotifier(api *slack.WebAPI, warningTimeout time.Duration) *SlackIMNotifier {
 	return &SlackIMNotifier{
-		im:    slack.NewInstantMessenger(api),
-		users: slack.NewTeamDirectory(api),
+		im:             slack.NewInstantMessenger(api),
+		users:          slack.NewTeamDirectory(api),
+		warningTimeout: warningTimeout,
 	}
 }
 
-func (notifier *SlackIMNotifier) DeployStarted(_ string, _ deploy.Deploy) {}
+func (notifier *SlackIMNotifier) DeployStarted(_ string, d deploy.Deploy) {
+	notifier.mutex.Lock()
+	defer notifier.mutex.Unlock()
+
+	notifier.timer = time.AfterFunc(notifier.warningTimeout, func() {
+		message := slack.Message{
+			Text: fmt.Sprintf("Your deploy %q was started %s ago. Are you sure you are still deploying", d.Subject, notifier.warningTimeout),
```

Comment:
```suggestion
			Text: fmt.Sprintf("Your deploy %q was started %s ago. Are you still deploying?", d.Subject, notifier.warningTimeout),
```
Maybe this can be worded more neutral. There's a chance a real deploy can take multiple hours.

On that note, maybe instead of a `time.AfterFunc()` we should use a ticker to check back every two hours? So if my deploy does take three hours I'd ignore the first notification, but be reminded to finish it when I see the second.

Reply:
at beginning I thought of ticker, but I am not sure if that's needed or might be too annoying?

Reply:
It's probably fine as is. If later we feel like it should check back again, we can consider switching to a ticker later 👍